File: IDD Dialer/DialingCodesHelper/DiallingCodesHelper.m
//
//  DiallingCodesHelper.m
//  IDD Dialer
//
//  Created by Raymond Lee on 4/3/14.
//  Copyright (c) 2014 RayCom. All rights reserved.
//

#import "DiallingCodesHelper.h"

@implementation DiallingCodesHelper
static DiallingCodesHelper *_helper;

+ (DiallingCodesHelper *)sharedHelper {
    if (!_helper) {
        _helper = [[DiallingCodesHelper alloc] init];
    }
    return _helper;
}

- (instancetype)init {
    self.iddArray = [DiallingCodesHelper initialIDDs];
    self.countryCodeArray = [DiallingCodesHelper initialCountryCodes];
    self.disabledCountryCodeArray = [DiallingCodesHelper initialDisabledCountryCodes:self.countryCodeArray];
    self.preferenceDict = [DiallingCodesHelper initialPreference];
    return [super init];
}

#pragma mark - country codes

+ (NSArray *)countryNames {
    static NSArray *_countryNames = nil;
    if (!_countryNames) {
        _countryNames = [[[[self countryNamesByCode] allValues] sortedArrayUsingSelector:@selector(caseInsensitiveCompare:)] copy];
    }
    return _countryNames;
}

+ (NSArray *)countryCodes {
    static NSArray *_countryCodes = nil;
    if (!_countryCodes) {
        _countryCodes = [[[self countryCodesByName] objectsForKeys:[self countryNames] notFoundMarker:@""] copy];
    }
    return _countryCodes;
}

+ (NSDictionary *)countryNamesByCode {
    static NSDictionary *_countryNamesByCode = nil;
    if (!_countryNamesByCode) {
        NSMutableDictionary *namesByCode = [NSMutableDictionary dictionary];
        for (NSString *code in [NSLocale ISOCountryCodes]) {
            NSString *identifier = [NSLocale localeIdentifierFromComponents:@{NSLocaleCountryCode : code}];
            NSString *countryName = [[NSLocale localeWithLocaleIdentifier:[NSLocale preferredLanguages][0]] displayNameForKey:NSLocaleIdentifier value:identifier];
            if (countryName) namesByCode[code] = countryName;
        }
        _countryNamesByCode = [namesByCode copy];
    }
    return _countryNamesByCode;
}

+ (NSDictionary *)countryCodesByName {
    static NSDictionary *_countryCodesByName = nil;
    if (!_countryCodesByName) {
        NSDictionary *countryNamesByCode = [self countryNamesByCode];
        NSMutableDictionary *codesByName = [NSMutableDictionary dictionary];
        for (NSString *code in countryNamesByCode) {
            codesByName[countryNamesByCode[code]] = code;
        }
        _countryCodesByName = [codesByName copy];
    }
    return _countryCodesByName;
}

+ (NSDictionary *)diallingCodesByCode {
    static NSDictionary *_diallingCodesByCodeDict;
    if (!_diallingCodesByCodeDict) {
        NSString *path = [[NSBundle mainBundle] pathForResource:
                          @"DiallingCodes"                         ofType:@"plist"];
        NSDictionary *diallingCodesDict = [[NSDictionary alloc] initWithContentsOfFile:path];
        _diallingCodesByCodeDict = [diallingCodesDict copy];
    }
    return _diallingCodesByCodeDict;
}

#pragma mark - init data

+ (NSMutableArray *)initialIDDs {
    static NSString *fileName = @"idd_record.plist";
    NSString *path = [[self documentsDirectory] stringByAppendingPathComponent:fileName];
    NSMutableArray *iddArray = [NSMutableArray arrayWithContentsOfFile:path];;
    //write default
    if (!iddArray || [iddArray count] == 0) {
        NSString *path = [[NSBundle mainBundle] pathForResource:
                          @"IDDData"                              ofType:@"plist"];
        
        iddArray = [[NSMutableArray alloc] initWithContentsOfFile:path];
        NSLog(@"INIT: %@ not exsist, creat default with size = %lu", fileName, iddArray.count);
    }else{
        NSLog(@"INIT: idd code = %lu", (unsigned long) iddArray.count);
    }
    return iddArray;
}

+ (NSMutableArray *)initialCountryCodes {
    static NSString *fileName = @"cc_record.plist";
    NSString *path = [[self documentsDirectory] stringByAppendingPathComponent:fileName];
    NSMutableArray *countryArray = [NSMutableArray arrayWithContentsOfFile:path];
    
    //write default
    if (!countryArray || [countryArray count] == 0) {
        path = [[NSBundle mainBundle] pathForResource:
                @"CountryCodeData"            ofType:@"plist"];
        countryArray = [[NSMutableArray alloc] initWithContentsOfFile:path];
        NSLog(@"INIT: %@ not exsist, creat default with size = %lu",fileName, countryArray.count);
    }else{
        NSLog(@"INIT: enabled country = %lu", (unsigned long) countryArray.count);
    }
    return countryArray;
}

+ (NSMutableArray *)initialDisabledCountryCodes:(NSArray*) countryArray {
    static NSString *fileName = @"dcc_record.plist";
    NSString *path = [[self documentsDirectory] stringByAppendingPathComponent:fileName];
    NSMutableArray *dCountryArray = [NSMutableArray arrayWithContentsOfFile:path];
    
    if (!dCountryArray || [dCountryArray count] == 0) {
        dCountryArray = [NSMutableArray new];
        for (NSString *code in [self countryCodes]) {
            if (![countryArray containsObject:code]) {
                [dCountryArray addObject:code];
            }
        }
        NSLog(@"INIT: %@ not exsist, creat default with size = %lu",fileName, dCountryArray.count);
    }else{
        NSLog(@"INIT: disabled country = %lu", (unsigned long) dCountryArray.count);
    }
    return dCountryArray;
}

+ (NSMutableDictionary *)initialPreference {
    static NSString *fileName = @"preference.plist";
    NSString *path = [[self documentsDirectory] stringByAppendingPathComponent:fileName];
    NSMutableDictionary *preDict = [NSMutableDictionary dictionaryWithContentsOfFile:path];
    if (!preDict) {
        preDict = [NSMutableDictionary dictionaryWithCapacity:0];
        NSLog(@"INIT: %@ not exsist",fileName);
    }else{
        NSLog(@"INIT: preference = %lu", (unsigned long) preDict.count);
    }
    return preDict;
}

+ (NSString *)countryNameByCode:(NSString *)code {
    return [self countryNamesByCode][code];
}

+ (NSString *)diallingCodeByCode:(NSString *)code {
    return [self diallingCodesByCode][code];
}


+ (NSString *)preferenceByCode:(NSString *)code {
    @synchronized (preferenceIDDs) {
        NSString *preference;
        preference = preferenceIDDs[code];
        if (preference) {
            NSInteger index = NSNotFound;
            for (NSDictionary *dict in idds) {
                if ([preference isEqual:dict[@"IDD"]]) {
                    index = [idds indexOfObject:dict];
                    break;
                }
            }
            if (index != NSNotFound) {
                return preference;
            } else {
                [self setPreference:nil code:code];
                
            }
        }
    }
    return nil;
}

+ (void)setPreference:(NSString *)preference code:(NSString *)code {
    if (code) {
        @synchronized (preferenceIDDs) {
            if (preference) {
                preferenceIDDs[code] = preference;
            } else {
                [preferenceIDDs removeObjectForKey:code];
            }
            NSString *path = [[self documentsDirectory] stringByAppendingPathComponent:@"preference.plist"];
            [preferenceIDDs writeToFile:path atomically:YES];
            NSLog(@"preference saved");
        }
    } else {
        BLog(@"code is nil");
    }
}

#pragma mark - others

+ (NSString *)documentsDirectory {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = paths[0];
    return documentsDirectory;
}

@end

File: IDD Dialer/DialingCodesHelper/DiallingCodesHelper.swift
//
//  DiallingCodesHelper.swift
//  IDD Dialer
//
//  Created by Raymond Lee on 18/6/14.
//  Copyright (c) 2014 RayCom. All rights reserved.
//

@objc class IDDRecord{
    var iddCode:String
    var with00:Bool
    
    init(iddCode:String, with00:Bool){
        self.iddCode = iddCode
        self.with00 = with00
    }
}

@infix func == (left: IDDRecord, right: IDDRecord) -> Bool {
    if left.iddCode == right.iddCode{
        return true
    }else{return false}
}

@objc class DiallingCodesHelper {
    var iddArray:IDDRecord[] = []
    var countryCodeArray:String[] = []
    var disabledCountryCodeArray:String[] = []
    var preferenceDict = Dictionary<String, String>()
    
    var countryNamesByCode:Dictionary<String, String>{
        struct Static {
            static var instance : Dictionary<String, String>? = nil
            static var token : dispatch_once_t = 0
            }
            dispatch_once(&Static.token) {
                var countryDict = Dictionary<String, String>()
            for code : AnyObject in NSLocale.ISOCountryCodes() {
                let identifier = NSLocale.localeIdentifierFromComponents([NSLocaleCountryCode : code])
                let countryName = NSLocale(localeIdentifier: (NSLocale.preferredLanguages() as NSArray).objectAtIndex(0) as String).displayNameForKey(NSLocaleIdentifier, value: identifier)
                countryDict[code as String] = countryName
                Static.instance = countryDict
            }
        }
        return Static.instance!
    }
    
    var countryCodesByName:Dictionary<String, String>{
    struct Static {
        static var instance : Dictionary<String, String>? = nil
        static var token : dispatch_once_t = 0
        }
        dispatch_once(&Static.token) {
            var _countryCodesByName = Dictionary<String, String>()
            for (code,name) in self.countryNamesByCode {
                _countryCodesByName[name] = code;
            }
            Static.instance = _countryCodesByName
        }
        return Static.instance!
    }
    
    var countryNames:String[]{
    struct Static {
        static var instance : String[]? = nil
        static var token : dispatch_once_t = 0
        }
        dispatch_once(&Static.token) {
            var _countryNames : String[]
            _countryNames = sort(Array(self.countryNamesByCode.values), {(s1:String, s2:String) -> Bool in return s1.lowercaseString < s2.lowercaseString})
            Static.instance = _countryNames
        }
        return Static.instance!
    }
    
    var countryCodes:String[]{
    struct Static {
        static var instance : String[]? = nil
        static var token : dispatch_once_t = 0
        }
        dispatch_once(&Static.token) {
            var _countryCodes = String[]()
            for name in self.countryNames{
                if let code = self.countryCodesByName[name]{
                    _countryCodes.append(code)
                }
            }
            Static.instance = _countryCodes
        }
        return Static.instance!
    }
    
    var diallingCodesByCode:Dictionary<String, String>{
    struct Static {
        static var instance : Dictionary<String, String>? = nil
        static var token : dispatch_once_t = 0
        }
        dispatch_once(&Static.token) {
            var _diallingCodesByCodeDict = Dictionary<String, String>()
            let path = NSBundle.mainBundle().pathForResource("DiallingCodes", ofType: "plist")
            _diallingCodesByCodeDict = NSDictionary(contentsOfFile: path) as Dictionary<String, String>
            Static.instance = _diallingCodesByCodeDict
        }
        return Static.instance!
    }
    
    class var sharedHelper : DiallingCodesHelper {
        get {
            struct Static {
                static var instance : DiallingCodesHelper? = nil
                static var token : dispatch_once_t = 0
            }
            
            dispatch_once(&Static.token) { Static.instance = DiallingCodesHelper() }
            
            return Static.instance!
    }
    }
    
    class func sharedHelperMethod() -> DiallingCodesHelper{
        return sharedHelper
    }
    
    init(){
        iddArray = initialIDDs()
        countryCodeArray = initialCountryCodes()
        disabledCountryCodeArray = initialDisabledCountryCodes()
        preferenceDict = initialPreference()
    }
    
    //    #pragma mark - convert data
    func convertToIDDRecords(array:Dictionary<String, AnyObject>[]) -> IDDRecord[]{
        var iddRecords = IDDRecord[]()
        for dict in array{
            var code = ""
            var with00 = false
            for (key, value : AnyObject) in dict{
                if key == "IDD"{
                    code = (value as String)
                }else if key == "IDD00"{
                    with00 = (value as NSNumber).boolValue
                }
            }
            iddRecords.append(IDDRecord(iddCode: code, with00: with00))
        }
        return iddRecords
    }
    
    func convertToDictionaries(iddRecords:IDDRecord[]) -> Dictionary<String, AnyObject>[]{
        var dictionaries = Dictionary<String, AnyObject>[]()
        for iddRecord in iddRecords{
            dictionaries.append(["IDD":iddRecord.iddCode, "IDD00":NSNumber(bool: iddRecord.with00)])
        }
        return dictionaries
    }
    
    //    #pragma mark - init data
    
    func initialIDDs() -> IDDRecord[]{
        var path = DiallingCodesHelper.documentsDirectory().stringByAppendingPathComponent("idd_record.plist")
        var tempIddArray = NSArray(contentsOfFile: path)
        //write default
        if tempIddArray.count == 0 {
            path = NSBundle.mainBundle().pathForResource("IDDData", ofType: "plist")
            tempIddArray = NSArray(contentsOfFile: path)
            NSLog("INIT: idd_record.plist not exsist, creat default with size = %lu", tempIddArray.count)
        }else{
            NSLog("INIT: idd = %lu", tempIddArray.count)
        }
        return convertToIDDRecords(tempIddArray as Dictionary<String, AnyObject>[]);
    }
    
    func initialCountryCodes() -> String[]{
        var path = DiallingCodesHelper.documentsDirectory().stringByAppendingPathComponent("cc_record.plist")
        var tempCountryArray = NSArray(contentsOfFile: path)
        //write default
        if tempCountryArray.count == 0 {
            path = NSBundle.mainBundle().pathForResource("CountryCodeData", ofType: "plist")
            tempCountryArray = NSArray(contentsOfFile: path)
            NSLog("INIT: cc_record.plist not exsist, creat default with size = %lu", tempCountryArray.count)
        }else{
            NSLog("INIT: enabled country = %lu", tempCountryArray.count)
        }
        return tempCountryArray as String[]
    }
    
    func initialDisabledCountryCodes() -> String[]{
        let path = DiallingCodesHelper.documentsDirectory().stringByAppendingPathComponent("dcc_record.plist")
        var dCountryArray = NSArray(contentsOfFile: path)
        //write default
        if dCountryArray.count == 0 {
            var dCountries = String[]()
            for code in countryCodes{
                if (!contains(countryCodeArray, code)) {
                    dCountries.append(code)
                }
            }
            NSLog("INIT: dcc_record.plist not exsist, creat default with size = %lu", dCountries.count)
            return dCountries
        }else{
            NSLog("INIT: disabled country = %lu", dCountryArray.count)            
        }
        return dCountryArray as String[]
    }
    
    func initialPreference() -> Dictionary<String, String>{
        let path = DiallingCodesHelper.documentsDirectory().stringByAppendingPathComponent("preference.plist")
        var preDict = NSDictionary(contentsOfFile: path)
        if preDict.count == 0 {
            preDict = NSDictionary()
            NSLog("INIT: preference.plist not exsist")
        }else{
            NSLog("INIT: preference = %lu", preDict.count)
        }
        return preDict as Dictionary<String, String>
    }
    
    func countryNameByCode(code:String) -> String?{
        return countryNamesByCode[code]
    }
    
    func diallingCodeByCode(code:String) -> String?{
        return diallingCodesByCode[code]
    }
    
    func preferenceByCode(code:String) -> String?{
        let preference = preferenceDict[code]
        var index = -1
        for i in 0..iddArray.count {
            if iddArray[i].iddCode == preference {
                index = i
                break
            }
        }
        if index != -1 {
            return preference;
        } else {
            setPreference(nil, code: code)
            return nil;
        }
    }
    
    func setPreference(preference:String?, code:String){
        if preference {
            preferenceDict[code] = preference
        } else {
            preferenceDict.removeValueForKey(code)
        }
        let path = DiallingCodesHelper.documentsDirectory().stringByAppendingPathComponent("preference.plist")
        preferenceDict.bridgeToObjectiveC().writeToFile(path, atomically: true)
    }
    
    
    class func documentsDirectory() -> String{
        struct Static {
            static var instance : String? = nil
            static var token : dispatch_once_t = 0
        }
        dispatch_once(&Static.token) {
        let paths = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)
            Static.instance = paths[0] as? String
        }
        return Static.instance!
    }
}

